Trello Node Affinity and New Section for Advanced Scheduling

mburke5678 opened the pull request.

mburke5678 commented:
@sjenning Is there someone who can take a look at this new section? 
Covers:
https://trello.com/c/sbQ6biAw/532-document-scheduler-custom-scheduler
https://trello.com/c/HxUMi5Ou/530-document-scheduler-node-affinity-anti-affinity
https://trello.com/c/ZBu10DF5/531-document-scheduler-pod-affinity-anti-affinity

Added new section on Advanced Scheduling to match Kubetnetes organization. Section has new topics on custom scheduler, node affinity, and pod affinity. Moved Taints and Tolerations into this section. Created new topic for node selector using an include from the existing Scheduler doc. 
See PR:
https://github.com/openshift/openshift-docs/pull/4597

mburke5678 commented:
@mdshuai moved the associated Trello related to Taints and Tolerations to Accepted. I closed the PR associated with the Taints and Tolerations topic https://github.com/openshift/openshift-docs/pull/4443 and will use this PR to push to publishing.

aveshagarwal commented on admin_guide/advanced_scheduling/custom_scheduler.adoc:
```diff
@@ -0,0 +1,99 @@
+
+. Execute the following command to check the status of the pod:
++
+$ oc get pod
```
`----` missing

aveshagarwal commented on admin_guide/advanced_scheduling/custom_scheduler.adoc:
```diff
@@ -0,0 +1,99 @@
+----
+$ oc describe pod custom-scheduler
+[...]
+Events:
```
formatting seems off here.

aveshagarwal commented on admin_guide/advanced_scheduling/index.adoc:
```diff
@@ -0,0 +1,27 @@
+Generally, advanced scheduling  is not necessary, as the {product-title} automatically places pods in a reasonable manner. For example, the default scheduler attempts to distributed pods across the nodes eveny and condieres the free resources in a node. 
+
+For some reason, you might want more control over where a pod is placed. For example a pod might need machine with an SSD attached, a pod needs to be on a machine with faster disks speed (or prevented from being placed on that machine), or pods from two different services need to be located so they can  communicate. 
+
```
extra space.

aveshagarwal commented on admin_guide/advanced_scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,153 @@
+
+== Overview
+
+_Node affinity_ is a set rules used by the scheduler to determine which nodes are selected for a pod. The rules are defined using custom xref:../architecture/core_concepts/pods_and_services.html#labels[labels on nodes] and selectors specified in pods. Rules can be either required or preferred, depending on how strictly you want the rules enforced. 
```
set of rules?

aveshagarwal commented on admin_guide/advanced_scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,153 @@
+            - e2e-az2
+----
+
+. Specify an `operator`. The xref:../../rest_api/openshift_v1.html#unversioned-labelselectorrequirement[operator can be `In`, `NotIn`, `Exists`, or `DoesNotExist`]. For example, use the operator `In` to require the label to be in the node.
```
Node affinity supports Gt and Lt too.

aveshagarwal commented on admin_guide/advanced_scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,153 @@
+
+[NOTE]
+====
+If labels on a node change at runtime that results in an node affinity rule on a pod no longer being met, the pod will still continue to run on the node. 
```
Would be good to explicit state that it is true for requiredDuringSchedulingIgnoredDuringExecution and preferredDuringSchedulingIgnoredDuringExecution. Other variants *requiredDuringExecution is not yet implemented.

aveshagarwal commented on admin_guide/advanced_scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,153 @@
+. Specify a weight for the node, 1-100. The node that with highest weight is preferred.
+
+. Specify an `operator`. The xref:../../rest_api/openshift_v1.html#unversioned-labelselectorrequirement[operator can be `In`, `NotIn`, `Exists`, or `DoesNotExist`]. For example, use the operator `In` to require the label to be in the node.
+
```
Gt and Lt too.

aveshagarwal commented on admin_guide/advanced_scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,153 @@
+
+* If you specify multiple `nodeSelectorTerms` associated with `nodeAffinity` types, then the pod can be scheduled onto a node if one of the `nodeSelectorTerms` is satisfied.
+
+* If you specify multiple `matchExpressions` associated with `nodeSelectorTerms`, then the pod can be scheduled onto a node only if all `matchExpressions` are be satisfied.
```
`are be`?

aveshagarwal commented on admin_guide/advanced_scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,150 @@
+
+== Overview
+
+_Pod affinity_ and _pod anti-affinity_ allows you to specify rules about how pods should be placed relative to one another. For example, you could spread or pack pods within a service or relative to pods in other services; or, you can make sure a pod is scheduled on a machine with an SSD on it. 
```
`you can make sure a pod is scheduled on a machine with an SSD on it.` does not seem to fit for pod affinities or anti-affinities.

aveshagarwal commented on admin_guide/advanced_scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,150 @@
+Pod affinity and anti-affinity allow you to constrain which nodes your pod is eligible to be scheduled on based on the labels on other pods. A xref:../../architecture/core_concepts/pods_and_services.html#labels[label] is a key/value pair.
+
+* Pod affinity tells the scheduler to locate a new pod on the same node as pods with the same labels.
+* Pod anti-affinity prevents the scheduler from locating a new pod on the same node as pods with the same labels.
```
`same labels` does not sound correct. Basically it is determined based on the label selectors in rules part of input pod and the labels on the pods on nodes.

aveshagarwal commented on admin_guide/advanced_scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,150 @@
+
+[NOTE]
+====
+If labels on a node change at runtime such that the affinity rules on a pod are no longer met, the pod will still continue to run on the node. 
```
As above only true for *ignoredDuringExecution, not for *requiredDuringExecution (not yet implemented).

mburke5678 replied:
@aveshagarwal Do you mean that the NOTE applies only to `requiredDuringSchedulingIgnoredDuringExecution` and `preferredDuringSchedulingIgnoredDuringExecution`. But not `requiredDuringSchedulingRequiredDuringExecution`, which is not documented here. So, no change is required here?

aveshagarwal replied:
I think yeah sounds good.

aveshagarwal commented on admin_guide/advanced_scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,428 @@
+[[admin-guide-taints]]
+= Taints and Tolerations
```
If I remember correctly, I had already reviewed this doc previously?

mburke5678 replied:
Correct.

aveshagarwal commented:
@mburke5678 some minor comments. Please let me know if any comment is not clear. Thanks.

mburke5678 commented:
@aveshagarwal I added some Examples to the node and pod affinity topics, taken from the test cases. Would you mind taking a quick look?

aveshagarwal commented:
@mburke5678 LGTM.

wmengRH commented on admin_guide/advanced_scheduling/custom_scheduler.adoc:
```diff
@@ -0,0 +1,100 @@
+----
+$ oc get pod
+NAME                READY     STATUS    RESTARTS   AGE
+custom-scheduler    0/1       Pending    0                    2m
```
move "2m" under "AGE"

wmengRH commented on admin_guide/advanced_scheduling/custom_scheduler.adoc:
```diff
@@ -0,0 +1,100 @@
+----
+$ oc get pod
+NAME                READY     STATUS    RESTARTS   AGE
+custom-scheduler    1/1       Running    0                   4m
```
move "4m"  under "AGE"

wmengRH commented on admin_guide/advanced_scheduling/custom_scheduler.adoc:
```diff
@@ -0,0 +1,100 @@
+Events:
+  FirstSeen  LastSeen  Count  From              SubObjectPath  Type       Reason Message
+  ---------  --------  -----  ----              -------------  --------   ------ -------
+  1m         1m        1      custom-scheduler  Normal         Scheduled  Successfully assigned custom-scheduler to <$node1>
```
"custom-scheduler" should be "my-scheduler" under "From"

mburke5678 commented:
@openshift/team-documentation PTAL at the  admin_guide/advanced_scheduling/custom_scheduler.adoc topic.

wmengRH commented on admin_guide/advanced_scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,401 @@
+!`Equal`
+!The `key`/`value`/`effect` parameters must match. This is the default.
+!`Exists`
+!The `key`/`effect` parameters must match. You can leave a blank `value` parameter, which will match any.
```
no value should be specified if the operator is Exists
it is a must, not can.

wmengRH commented on admin_guide/advanced_scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,401 @@
+  operator: "Equal" <1>
+  value: "value" <1>
+  effect: "NoSchedule" <1>
+  tolerationSeconds: 3600 <2>
```
effect must be 'NoExecute' when `tolerationSeconds` is set

wmengRH commented on admin_guide/advanced_scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,401 @@
+- key: "key"
+  operator: "Exists"
+  effect: "NoSchedule"
+  tolerationSeconds: 3600
```
effect must be 'NoExecute' when `tolerationSeconds` is set

wmengRH commented on admin_guide/advanced_scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,401 @@
+----
+[source, yaml]
+tolerations:
+- key: "key"
```
- key: "key"
- key: "key1"

wmengRH commented on admin_guide/advanced_scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,401 @@
+----
+[source, yaml]
+tolerations:
+- key: "key" <1>
```
key: "key"
key: "key1"

wmengRH commented on admin_guide/advanced_scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,401 @@
+tolerations:
+- key: "key" <1>
+  operator: "Equal" <1>
+  value: "value" <1>
```
+  value: "value" <1>
+  value: "value1" <1>

wmengRH commented on admin_guide/advanced_scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,401 @@
+.. Check the pod tolerations
++
+----
+$ oc describe pod <pod-name> -n dma|grep -i toleration
```
suggest remove "-n dma"

wmengRH commented on admin_guide/advanced_scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,401 @@
+For example:
++
+----
+$ oc describe pod hello-pod -n dma|grep -i toleration
```
suggest remove "-n dma"

wmengRH commented on admin_guide/advanced_scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,401 @@
+== Daemonsets and Tolerations
+
+link:https://kubernetes.io/docs/admin/daemons/[DaemonSet] pods are created with `NoExecute` tolerations for `node.alpha.kubernetes.io/unreachable` and `node.alpha.kubernetes.io/notReady`
+with no `tolerationSeconds` to ensure that DaemonSet pods are never evicted due to these problems, even when when the Default Toleration Seconds feature is disabled.
```
remove duplicated "when when"

wmengRH commented on admin_guide/advanced_scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,401 @@
+For example:
++
+----
+$ oadm taint nodes node1 dedicated=groupName:NoSchedule`
```
remove extra ` at the end

wmengRH commented on admin_guide/advanced_scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,401 @@
+For example:
++
+----
+$ oadm taint nodes node1 dedicated=groupName:NoSchedule`
```
extra `  at the end

wmengRH commented on admin_guide/advanced_scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,401 @@
++
+----
+$ oadm taint nodes <node-name> special=true:NoSchedule
+$ oadm taint nodes <node-name> special=true:PreferNoSchedule
```
I am not suggesting use "true/false/yes/no" as value, since those string-like values will be translated / treated as "1/0" in pod yaml file.
I prefer disktype=ssd or something like this.

wmengRH commented on admin_guide/advanced_scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,283 @@
+====
+
+[[admin-guide-sched-affinity-config]]
+== Configuring Node Infinity
```
Node Infinity -> Node Affinity

wmengRH commented on admin_guide/advanced_scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,283 @@
+[[admin-guide-sched-affinity-config]]
+== Configuring Node Infinity
+
+Node affinity is configured through the pod specification file. You can specify a xref:admin-guide-sched-affinity-config-req[required rule], a xref:admin-guide-sched-affinity-config-pref[preferred rule], or both. If you specify both, the node must first meet the required rule, then meet the preferred.
```
then meet the preferred
not must, but just try best

wmengRH commented on admin_guide/advanced_scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,283 @@
+<1> The stanza to configure node affility.
+<2> Defines a required rule.
+<3> The key/value pair (label) that must be matched to apply the rule.
+<4> The xref:../../rest_api/openshift_v1.html#unversioned-labelselectorrequirement[operator] represents the relationship between the key and a set of values.  Can be `In`, `NotIn`, `Exists`, or `DoesNotExist`, `lt` or `gt`.
```
1. not the relationship between the key and a set of values.
it is the ralationship between ( (the value of the key) on the node) and (the set of values in matchexpressions).
2. it is case sensitive,  lt gt should be Lt Gt.

wmengRH commented on admin_guide/advanced_scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,283 @@
+      - weight: 1 <3>
+        preference:
+          matchExpressions:
+          - key: e2e-az-NorthSouth <4>
```
e2e-az-EastWest

wmengRH commented on admin_guide/advanced_scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,283 @@
+<2> Defines a preferred rule.
+<3> Specifies a weight for a preferred rule. The node that with highest weight is preferred.
+<4> The key/value pair (label) that must be matched to apply the rule.
+<5> The xref:../../rest_api/openshift_v1.html#unversioned-labelselectorrequirement[operator] represents the relationship between the key and a set of values.  Can be `In`, `NotIn`, `Exists`, or `DoesNotExist`, `lt` or `gt`.
```
need to modify, mentioned above.

wmengRH commented on admin_guide/advanced_scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,283 @@
++
+.. Specify the key and values that must be met. If you want the new pod to be scheduled on the node you edited, use the same `key` and `value` parameters as the label in the node.
++
+.. Specify an `operator`. The xref:../../rest_api/openshift_v1.html#unversioned-labelselectorrequirement[operator can be `In`, `NotIn`, `Exists`, `DoesNotExist`, `lt` or `gt`]. For example, use the operator `In` to require the label to be in the node.
```
lt -> Lt
gt -> Gt

wmengRH commented on admin_guide/advanced_scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,283 @@
+
+. Specify a weight for the node, 1-100. The node that with highest weight is preferred.
+
+. Specify an `operator`. The xref:../../rest_api/openshift_v1.html#unversioned-labelselectorrequirement[operator can be `In`, `NotIn`, `Exists`, `DoesNotExist`, `lt` or `gt`]. For example, use the operator `In` to require the label to be in the node.
```
lt -> Lt
gt -> Gt

wmengRH commented on admin_guide/advanced_scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,283 @@
+* The pod *pod-s1* cannot be scheduled on *Node1*.
++
+----
+oc describe pod node1
```
pod node1 -> pod pod-s1

wmengRH commented on admin_guide/advanced_scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,351 @@
+
+== Overview
+
+_Pod affinity_ and _pod anti-affinity_ allow you to specify rules about how pods should be placed relative to one another. For example, you could spread or pack pods within a service or relative to pods in other services or, you can make sure a pod is scheduled on a specific type of machine.
```
I do not agree with this "you can make sure a pod is scheduled on a specific type of machine." for pod affinity
correct me if I am wrong.

mburke5678 replied:
@wmengRH This statement came from the Kubernetes doc: 
"there are some circumstances where you may want more control on a node where a pod lands, e.g. to ensure that a pod ends up on a machine with an SSD attached to it, or to co-locate pods from two different services that communicate a lot into the same availability zone."
I made the "machine with an SSD attached" to a more general statement. 
Maybe I can say instead: "you can make sure a pod is scheduled on **a specific machine**."
Would that make more sense?

wmengRH replied:
wow, can you show me the link? I want to read the context and user case to understand what it is.
Thanks.

mburke5678 replied:
@wmengRH 
I am using this document. 
https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#per-pod-configurable-eviction-behavior-when-there-are-node-problems-alpha-feature

Please tell me if i am incorrect, which i probably am wrong!

wmengRH replied:
I found "e.g. to ensure that a pod ends up on a machine with an SSD attached to it, or to co-locate pods from two different services that communicate a lot into the same availability zone."
in the link.
"to ensure that a pod ends up on a machine with an SSD attached to it" is about node affinity.
"to co-locate pods from two different services that communicate a lot into the same availability zone" is about pod affinity.
so it is not the right place for it.

mburke5678 replied:
@wmengRH Thank you for clarifying. 
I took some examples from the use cases in the design doc. Is this more accurate?
"_Pod affinity_ and _pod anti-affinity_ allow you to specify rules about how pods should be placed relative to one another. For example, using affinity rules, you could spread or pack pods within a service or relative to pods in other services. Anti-affinity rules allow you to prevent pods of a particular service  from scheduling  on the same nodes as pods of another service that are known to interfere with the performance of the pods of the first service. Or, you could spread the pods of a service across nodes or availability zones to reduce correlated failures. "
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/podaffinity.md

mburke5678 replied:
For node affinity:
"Node affinity allows you to constrain which nodes your pod is eligible to be scheduled on using labels on the node. For example, you could configure a pod to only run on a node with a specific CPU or in a specific availability zone."
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/nodeaffinity.md

wmengRH replied:
Looks pretty good. Thank you.

wmengRH commented on admin_guide/advanced_scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,351 @@
+[[admin-guide-sched-affinity-pod-config]]
+== Configuring Pod Affinity/Anti-Affinity
+
+You configure pod affinity/anti-affinity through the pod specification files. You can specify a xref:admin-guide-sched-affinity-config-req[required rule], a xref:admin-guide-sched-affinity-config-pref[preferred rule], or both. If you specify both, the node must first meet the required rule, then meet the preferred.
```
then meet the preferred
not must, but just try best

wmengRH commented on admin_guide/advanced_scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,351 @@
+
+The following example shows a pod specification configured for pod affinity and anti-affinity.
+
+In this example, the pod affinity rule indicates that the pod can schedule onto a node only if that node has at least one already-running pod with a label that has the key `security` and value `S1`. The pod anti-affinity rule says that the pod prefers to not schedule onto a node if that node is already running a pod with label having key `security `and value `S2`.
```
I got this in view page
security `and value `S2.
so I guess
`security `and value `S2`.   ->
`security` and value `S2`.

wmengRH replied:
I am not good at markdown format, hope you understand what I mean

wmengRH commented on admin_guide/advanced_scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,351 @@
+<1> Stanza to configure pod affinity.
+<2> Defines a required rule.
+<3> The key and value (label) that must be matched to apply the rule.
+<4> The xref:../../rest_api/openshift_v1.html#unversioned-labelselectorrequirement[operator] represents the relationship between the key and a set of values.  Can be `In`, `NotIn`, `Exists`, or `DoesNotExist`.
```
not the relationship between the key and a set of values.
it is the ralationship between (( (the value of the key) of label) on pod) and (the set of values in matchexpressions).

mburke5678 replied:
@wmengRH Is this correct:
<4> The operator represents the relationship between the **label on the existing pod** and the set of **values in the `matchExpression` parameters in the specification for the new pod**.
In our docs label=key/value.

wmengRH replied:
sorry. my misleading. 
I think the operator is self-explained, no need to make a translation. 
and for Exists and DoseNotExist, do not use values at all. 
It looks fine for me if only tell user the supported operators.

wmengRH commented on admin_guide/advanced_scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,351 @@
+<2> Defines a preferred rule.
+<3> Specifies a weight for a preferred rule. The node that with highest weight is preferred.
+<4> The key and value (label) that must be matched to apply the rule.
+<5> The xref:../../rest_api/openshift_v1.html#unversioned-labelselectorrequirement[operator] represents the relationship between the key and a set of values.  Can be `In`, `NotIn`, `Exists`, or `DoesNotExist`.
```
not the relationship between the key and a set of values.
it is the ralationship between (( (the value of the key) of label) on pod) and (the set of values in matchexpressions).

wmengRH commented on admin_guide/advanced_scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,351 @@
+  containers:
+  - name: ocp
+    image: docker.io/ocpqe/hello-pod
+----
```
for matching step 3, you might want use this to replace team4.yaml
cat pod-s1.yaml
```
apiVersion: v1
kind: Pod
metadata:
  name: security-s1
  labels:
    security: S1
spec:
  containers:
  - name: security-s1
    image: docker.io/ocpqe/hello-pod
```

wmengRH commented on admin_guide/advanced_scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,351 @@
+  containers:
+  - name: ocp
+    image: docker.io/ocpqe/hello-pod
+----
```
for matching step 3, you might want use this to replace team4.yaml
cat pod-s2.yaml
```
apiVersion: v1
kind: Pod
metadata:
  name: security-s2
  labels:
    security: S2
spec:
  containers:
  - name: security-s2
    image: docker.io/ocpqe/hello-pod
```

wmengRH commented on admin_guide/advanced_scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,351 @@
+[[admin-guide-sched-affinity-config-pod-pref]]
+=== Configuring an Anti-affinity Rule
+
+The following steps demonstrate a simple two-pod configuration that creates pod with a label and a pod that uses anti-affinity to prevent scheduling with that pod.
```
in the example, preferred used, so here 
to prevent 
trying to prevent    (exactly)

wmengRH commented on admin_guide/advanced_scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,351 @@
+              operator: In
+              values:
+              - s1
+          topologyKey: kubernetes.io/hostname
```
line 292 to 297 need to remove 2 spaces in the front of each line

wmengRH commented on admin_guide/advanced_scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,351 @@
+    image: docker.io/ocpqe/hello-pod
+----
+
+* The pod *pod-s2* cannot be scheduled on the same node as `pod-s1`.
```
"The pod *pod-s2* cannot be scheduled on the same node as `pod-s1`." is incorrect. 
pod-s2 has no relation with pod-s1.
(it is possble that pod-s2 is scheduled to the node pod-s1 is running on if there is a pod with lable security=s2 on it)
pod-s2 will propably be pending due to no pod with lable security=s2 found.

NAME   READY     STATUS    RESTARTS   AGE       IP        NODE
pod-s2        0/1       Pending    0                    23s       <none>    

Events:
  FirstSeen	LastSeen	Count	From			SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----			-------------	--------	------			-------
  30s		15s		6	default-scheduler			Warning		FailedScheduling	No nodes are available that match all of the following predicates:: MatchInterPodAffinity (1).

mburke5678 replied:
"The pod *pod-s2* will not be scheduled unless there is a node with a pod that has a label with security=s2."

wmengRH replied:
you missed it. @mburke5678

mburke5678 commented:
@openshift/team-documentation  PTAL

This PR covers:
https://trello.com/c/sbQ6biAw/532-document-scheduler-custom-scheduler
https://trello.com/c/HxUMi5Ou/530-document-scheduler-node-affinity-anti-affinity
https://trello.com/c/ZBu10DF5/531-document-scheduler-pod-affinity-anti-affinity

NO NEED TO REVIEW:
**Taints and Tolerations** has been peer-reviewed in [4443](https://github.com/openshift/openshift-docs/pull/4443)
 **Node selector** is imported from existing docs using include::

I added Taints and Tolerations to this PR to match the Kubernetes docs. The Kubernetes docs contains customer scheduler, node affinity, pod affinity, taints, and node selector in the same topic (https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#taints-and-tolerations-beta-feature).

bfallonf commented on admin_guide/scheduling/custom_scheduler.adoc:
```diff
@@ -0,0 +1,109 @@
+
+== Overview
+
+{product-title} supports the use of custom (your own or third-party) pod schedulers alongside the default scheduler that comes installed. While the default scheduler should be sufficient for most situations, there might be times when you want to use a different scheduler. You can run multiple schedulers alongside the default scheduler and configure which scheduler to use for each pods. 
```
I feel like, out of all of the above, only the last sentence is relevant. I'd suggest removing all except that.

bfallonf replied:
Hmm maybe change the sentence tho. Something like "You can run multiple, custom schedulers alongside the default scheduler and configure which scheduler to use for each pods."

bfallonf commented on admin_guide/scheduling/custom_scheduler.adoc:
```diff
@@ -0,0 +1,109 @@
+
+{product-title} supports the use of custom (your own or third-party) pod schedulers alongside the default scheduler that comes installed. While the default scheduler should be sufficient for most situations, there might be times when you want to use a different scheduler. You can run multiple schedulers alongside the default scheduler and configure which scheduler to use for each pods. 
+
+In order to schedule a given pod using a specific scheduler, specify the name of the scheduler in that pod spec. 
```
Should this link to somewhere that says this?

bfallonf commented on admin_guide/scheduling/custom_scheduler.adoc:
```diff
@@ -0,0 +1,109 @@
+
+[NOTE]
+====
+Information on how to create and deploy a scheduler is outside the scope of this document. For an example, see link:https://github.com/kubernetes/kubernetes/tree/master/plugin/pkg/scheduler[*_plugin/pkg/scheduler_* in the Kubernetes source directory].
```
Maybe s/create and deploy/create.

bfallonf commented on admin_guide/scheduling/custom_scheduler.adoc:
```diff
@@ -0,0 +1,109 @@
+. Execute the following command to check that the pod was created with the custom scheduler: 
++
+----
+check pod spec
```
Are they supposed to run this? Or the just below? Is the above needed?

bfallonf commented on admin_guide/scheduling/custom_scheduler.adoc:
```diff
@@ -0,0 +1,109 @@
+    image: gcr.io/google_containers/pause:2.0
+----
+
+<1> The name of the scheduler to use. When no scheduler name is supplied, the pod is automatically scheduled using the default-scheduler.
```
Does default-scheduler need markup? Would it be the same to just say "default scheduler"?

mburke5678 replied:
I think it is OK to just say "default scheduler" with no markup. It is like "default" setting" or "default value" which appears to not take markup.

bfallonf commented on admin_guide/scheduling/index.adoc:
```diff
@@ -0,0 +1,35 @@
+
+In situations where you might want more control over where new pods are placed, the {product-title} advanced scheduling features allow you to configure a pod so that the pod is required to (or has a preference to) run on a particular node or alongside a specific pod. Advanced scheduling also allows you to prevent pods from being placed on a node or with another pod.  
+
+For information on all of the advanced scheduling tools, see xref:../../admin_guide/scheduling/scheduler-advanced.adoc#admin-guide-scheduling-advanced[Advanced Scheduling].
```
s/on/for or maybe "about" or maybe just "for more information..."

bfallonf commented on admin_guide/scheduling/index.adoc:
```diff
@@ -0,0 +1,35 @@
+Pod scheduling is an internal process that determines how new
+pods are placed onto nodes within the cluster
+
+== Default scheduling
```
Anchors for all the topic headings in this section.

bfallonf commented on admin_guide/scheduling/index.adoc:
```diff
@@ -0,0 +1,35 @@
+
+== Custom schedulers
+
+In addition, {product-title} allows you to use your own or thrid-party schedulers simply by editing the pod specification. 
```
s/thrid-party/third-party

bfallonf replied:
Also remove "simply". Adverbs are a no-no.

bfallonf commented on admin_guide/scheduling/index.adoc:
```diff
@@ -0,0 +1,35 @@
+
+In addition, {product-title} allows you to use your own or thrid-party schedulers simply by editing the pod specification. 
+
+For information, see xref:../../admin_guide/scheduling/custom_scheduler.adoc#admin-guide-sched-custom[Custom Schedulers].
```
"For _more_ information..."

bfallonf commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,281 @@
+
+_Node affinity_ is a set of rules used by the scheduler to determine which nodes are selected for a pod. The rules are defined using custom xref:../architecture/core_concepts/pods_and_services.adoc#labels[labels on nodes] and selectors specified in pods.
+
+Node affinity allows you to constrain which nodes your pod is eligible to be scheduled on using labels on the node. For example, you could configure a pod to only run on a node with a specific CPU or in a specific availability zone.
```
Is the first sentence saying anything different form the first paragraph? It seems redundant. the example is good though.

bfallonf commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,281 @@
+
+Node affinity allows you to constrain which nodes your pod is eligible to be scheduled on using labels on the node. For example, you could configure a pod to only run on a node with a specific CPU or in a specific availability zone.
+
+There are two types of node affinity rules, _required_ and _preferred_.
```
s/,/:

bfallonf commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,281 @@
+
+There are two types of node affinity rules, _required_ and _preferred_.
+
+Required rules *must* be met before a pod can be scheduled on a node. Preferred rules specify that, if the rule is met, the scheduler will try to enforce the rules, but will not guarantee enforcement.
```
s/but will not/but does not

bfallonf commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,281 @@
+<1> The stanza to configure node affinity.
+<2> Defines a required rule.
+<3> The key/value pair (label) that must be matched to apply the rule.
+<4> The xref:../../rest_api/openshift_v1.adoc#unversioned-labelselectorrequirement[operator] represents the relationship between the label on the node and the set of values in the `matchExpression` parameters in the pod specification.  Can be `In`, `NotIn`, `Exists`, or `DoesNotExist`, `Lt` or `Gt`.
```
Oxford comma after `Lt`

bfallonf commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,281 @@
+
+<1> The stanza to configure node affinity.
+<2> Defines a preferred rule.
+<3> Specifies a weight for a preferred rule. The node that with highest weight is preferred.
```
s/that with/with

bfallonf commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,281 @@
+<2> Defines a preferred rule.
+<3> Specifies a weight for a preferred rule. The node that with highest weight is preferred.
+<4> The key/value pair (label) that must be matched to apply the rule.
+<4> The xref:../../rest_api/openshift_v1.adoc#unversioned-labelselectorrequirement[operator] represents the relationship between the label on the node and the set of values in the `matchExpression` parameters in the pod specification.  Can be `In`, `NotIn`, `Exists`, or `DoesNotExist`, `Lt` or `Gt`.
```
Same Oxford comma comment from before

bfallonf commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,281 @@
++
+.. Specify the key and values that must be met. If you want the new pod to be scheduled on the node you edited, use the same `key` and `value` parameters as the label in the node.
++
+.. Specify an `operator`. The xref:../../rest_api/openshift_v1.adoc#unversioned-labelselectorrequirement[operator can be `In`, `NotIn`, `Exists`, `DoesNotExist`, `Lt` or `Gt`]. For example, use the operator `In` to require the label to be in the node:
```
Oxford comma

bfallonf commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,281 @@
+
+. In the pod specification, use the `nodeAffinity` stanza to configure the `preferredDuringSchedulingIgnoredDuringExecution` parameter:
++
+.. Specify a weight for the node, 1-100. The node that with highest weight is preferred.
```
s/that with/with

bfallonf commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,281 @@
+            - e2e-az3
+----
+
+. Specify an `operator`. The xref:../../rest_api/openshift_v1.adoc#unversioned-labelselectorrequirement[operator can be `In`, `NotIn`, `Exists`, `DoesNotExist`, `Lt` or `Gt`]. For example, use the operator `In` to require the label to be in the node.
```
Oxford comma

bfallonf commented on admin_guide/scheduling/node_selector.adoc:
```diff
@@ -0,0 +1,48 @@
+
+== Overview
+
+A _node selector_ is specifies a map of key-value pairs. The rules are defined using custom xref:../../architecture/core_concepts/pods_and_services.adoc#labels[labels on nodes] and selectors specified in pods.
```
s/is specifies/specifies

bfallonf commented on admin_guide/scheduling/node_selector.adoc:
```diff
@@ -0,0 +1,48 @@
+
+A _node selector_ is specifies a map of key-value pairs. The rules are defined using custom xref:../../architecture/core_concepts/pods_and_services.adoc#labels[labels on nodes] and selectors specified in pods.
+
+For the pod to be eligible to run on a node, the pod must have the indicated key-value pairs as the label in the node. 
```
s/label in the node/label on the node

bfallonf commented on admin_guide/scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,362 @@
+
+_Pod affinity_ and _pod anti-affinity_ allow you to specify rules about how pods should be placed relative to one another. For example, using affinity rules, you could spread or pack pods within a service or relative to pods in other services. Anti-affinity rules allow you to prevent pods of a particular service  from scheduling  on the same nodes as pods of another service that are known to inte ...
+
+Pod affinity/anti-affinity allow you to constrain which nodes your pod is eligible to be scheduled on based on the labels on other pods. A xref:../../architecture/core_concepts/pods_and_services.adoc#labels[label] is a key/value pair.
```
s/allow/allows

bfallonf commented on admin_guide/scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,362 @@
+
+Pod affinity/anti-affinity allow you to constrain which nodes your pod is eligible to be scheduled on based on the labels on other pods. A xref:../../architecture/core_concepts/pods_and_services.adoc#labels[label] is a key/value pair.
+
+* Pod affinity can tell the scheduler to locate a new pod on the same node as other pods if the label selector in the new pod matches the label in the current pod.
```
s/label selector in the/label selector on the

bfallonf replied:
This happens a lot. Maybe I'm wrong.

bfallonf commented on admin_guide/scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,362 @@
+* Pod affinity can tell the scheduler to locate a new pod on the same node as other pods if the label selector in the new pod matches the label in the current pod.
+* Pod anti-affinity can prevent the scheduler from locating a new pod on the same node as pods with the same labels if the label selector in the new pod matches the label in the current pod.
+
+There are currently two types of pod affinity rules, _required_ and _preferred_.
```
s/,/:

bfallonf commented on admin_guide/scheduling/scheduler-advanced.adoc:
```diff
@@ -0,0 +1,51 @@
+
+Advanced scheduling involves configuring a pod so that the pod is required to run on particular nodes or has a preference to run on particular nodes.
+
+Generally, advanced scheduling  is not necessary, as the {product-title} automatically places pods in a reasonable manner. For example, the default scheduler attempts to distribute pods across the nodes evenly and considers the free resources in a node. However, you might want more control over where a pod is placed.
```
A suggestion: s/free/available

I think "free" might come with "not having to pay money" connotations.

bfallonf commented on admin_guide/scheduling/scheduler-advanced.adoc:
```diff
@@ -0,0 +1,51 @@
+
+There are several ways to invoke advanced scheduling in your cluster:
+
+== Pod Affinity and Anti-Affinity 
```
Anchor above, and all the other ones in this topic.

bfallonf commented on admin_guide/scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,398 @@
+** the `effect` parameters are the same.
+
+[[discrete]]
+=== Multiple taints
```
Caps for taint

bfallonf commented on admin_guide/scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,398 @@
+$ oadm taint nodes node1 dedicated=groupName:NoSchedule
+----
+
+. Add a corresponding toleration to the pods. This would be done most easily by writing a custom xref:../install_config/master_node_configuration.adoc#master-config-admission-control-config[admission controller].
```
To avoid the "easily" adverb, maybe:

Add a corresponding toleration to the pods by writing a custom admission controller."

wmengRH commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -18,7 +18,7 @@ For example, you could configure a pod to only run on a node with a specific CPU
 There are two types of node affinity rules: _required_ and _preferred_.
 
-Required rules *must* be met before a pod can be scheduled on a node. Preferred rules specify that, if the rule is met, the scheduler will try to enforce the rules, but does not guarantee enforcement.
+Required rules *must* be met before a pod can be scheduled on a node. Preferred rules specify that, if the rule is met, the scheduler will try to enforce the rules, but does not guarantee enforcement. If a pod has an affinity rule for which there is no node with a matching label, that pod can be scheduled on a node that does not have a label 
```
The last sentence is confusing

mburke5678 replied:
@wmengRH 
"If a pod has an affinity rule, but there is no node with a matching label, that pod can be scheduled on a node that does not have a label selector. "

wmengRH replied:
@mburke5678  
when no node meet the node affinity rule,
if it is required, pod is not scheduled, in pending, 
if it is preferred, all schedulable nodes are treated equally from node affinity point.

gaurav-nelson commented on admin_guide/scheduler.adoc:
```diff
@@ -287,6 +288,13 @@ of applications by allowing administrators to ensure that peer pods do not end
 up being too geographically separated. If no node is available within the same
 affinity group to host the pod, then the pod will not get scheduled.
 
+If you need greater control over where pods are scheduled, see 
```
s/where pods/where _the_ pods

gaurav-nelson replied:
@mburke5678 Can you revisit this one? Please also check all instances of the words `will` & `in order to` and use it only where it is necessary. Rest is all good. This was a big one indeed. :clap:

gaurav-nelson commented on admin_guide/scheduling/custom_scheduler.adoc:
```diff
@@ -0,0 +1,108 @@
+
+You can run multiple, custom schedulers alongside the default scheduler and configure which scheduler to use for each pods.
+
+In order to schedule a given pod using a specific scheduler, xref:admin-guide-sched-custom-deploy[specify the name of the scheduler in that pod spec]. 
```
s/In order to/ To

gaurav-nelson commented on admin_guide/scheduling/custom_scheduler.adoc:
```diff
@@ -0,0 +1,108 @@
+
+<1> The name of the scheduler to use. When no scheduler name is supplied, the pod is automatically scheduled using the default scheduler.
+
+. Execute the following command to create the pod:
```
Replace "Execute" with "Run" or similar. (Multiple instances)
IBM: In technical content, use only if a simpler, less technical term, such as "run" or "issue," or a more specific term, such as "start," "enter," or "open" is inappropriate.

gaurav-nelson commented on admin_guide/scheduling/custom_scheduler.adoc:
```diff
@@ -0,0 +1,108 @@
+The name of the scheduler is listed.
++
+----
+[...]
```
I am not sure about this one. Does this mean that the output shown here is truncated?

mburke5678 replied:
Yes. I believe so.

gaurav-nelson commented on admin_guide/scheduling/index.adoc:
```diff
@@ -0,0 +1,38 @@
+== Overview
+
+Pod scheduling is an internal process that determines how new
+pods are placed onto nodes within the cluster
```
missing full stop.

gaurav-nelson commented on admin_guide/scheduling/index.adoc:
```diff
@@ -0,0 +1,38 @@
+[[admin-guide-scheduling-index-default]]
+== Default scheduling
+
+{product-title} comes with a default scheduler that should serve the needs of most users. The default scheduler uses both inherent and customizable tools to determine the best fit for a pod. 
```
s/should serve/ serves

gaurav-nelson commented on admin_guide/scheduling/index.adoc:
```diff
@@ -0,0 +1,38 @@
+
+{product-title} comes with a default scheduler that should serve the needs of most users. The default scheduler uses both inherent and customizable tools to determine the best fit for a pod. 
+
+For information on how the default scheduler determines pod placement and on customizable parameters, see xref:../../admin_guide/scheduling/scheduler.adoc#admin-guide-scheduler[Default Scheduling].
```
Maybe: For information on how the default scheduler determines pod placement and _available_ customizable parameters,

gaurav-nelson commented on admin_guide/scheduling/index.adoc:
```diff
@@ -0,0 +1,38 @@
+[[admin-guide-scheduling-index-custom]]
+== Custom schedulers
+
+In addition, {product-title} allows you to use your own or third-party schedulers simply by editing the pod specification. 
```
{product-title} _also_ allows you to use your own or third-party schedulers by editing the pod specification. (added also, removed In addition, and simply)

gaurav-nelson commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,288 @@
+
+== Overview
+
+_Node affinity_ is a set of rules used by the scheduler to determine where a pod can be placed. The rules are defined using custom xref:../architecture/core_concepts/pods_and_services.adoc#labels[labels on nodes] and label selectors specified in pods. Node affinity allows a *pod* to specify an affinity (or anti-affinity) towards a group of *nodes* it can be placed on.  The node does not have cont ...
```
Remove extra space between "it can be placed on." and "The node does"

gaurav-nelson commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,288 @@
+
+There are two types of node affinity rules: _required_ and _preferred_.
+
+Required rules *must* be met before a pod can be scheduled on a node. Preferred rules specify that, if the rule is met, the scheduler will try to enforce the rules, but does not guarantee enforcement. If a pod has an affinity rule, but there is no node with a matching label, that pod can be scheduled on a node that does not have a label selector. 
```
s/scheduler will try/scheduler tries

gaurav-nelson commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,288 @@
+
+[NOTE]
+====
+If labels on a node change at runtime that results in an node affinity rule on a pod no longer being met, the pod will still continue to run on the node.
```
s/will still continue/continues

gaurav-nelson commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,288 @@
+[[admin-guide-sched-affinity-config]]
+== Configuring Node Affinity
+
+Node affinity is configured through the pod specification file. You can specify a xref:admin-guide-sched-affinity-config-req[required rule], a xref:admin-guide-sched-affinity-config-pref[preferred rule], or both. If you specify both, the node must first meet the required rule, then will attempt to meet the preferred rule.
```
s/then will attempt to meet the preferred rule./then _it attempts_ to meet the preferred rule.

gaurav-nelson commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,288 @@
+<1> The stanza to configure node affinity.
+<2> Defines a required rule.
+<3> The key/value pair (label) that must be matched to apply the rule.
+<4> The xref:../../rest_api/openshift_v1.adoc#unversioned-labelselectorrequirement[operator] represents the relationship between the label on the node and the set of values in the `matchExpression` parameters in the pod specification.  Can be `In`, `NotIn`, `Exists`, or `DoesNotExist`, `Lt`, or `Gt`.
```
remove extra spaces between "specification." and "Can"

gaurav-nelson commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,288 @@
+<2> Defines a preferred rule.
+<3> Specifies a weight for a preferred rule. The node with highest weight is preferred.
+<4> The key/value pair (label) that must be matched to apply the rule.
+<4> The xref:../../rest_api/openshift_v1.adoc#unversioned-labelselectorrequirement[operator] represents the relationship between the label on the node and the set of values in the `matchExpression` parameters in the pod specification.  Can be `In`, `NotIn`, `Exists`, or `DoesNotExist`, `Lt`, or `Gt`.
```
Remove extra space before "Can"

gaurav-nelson commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,288 @@
+
+The following steps demonstrate a simple configuration that creates a node and a pod that the scheduler is required to place on the node.
+
+. Add a label to a node by editing the node configuration or by executing the `oc label node` command:
```
s/executing/running or something similar

gaurav-nelson commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,288 @@
+[[admin-guide-sched-affinity-config-pref]]
+=== Configuring a Preferred Rule
+
+The following steps demonstrate a simple configuration that creates a node and a pod that the scheduler will try to place on the node.
```
s/will try/tries

gaurav-nelson commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,288 @@
+
+. In the pod specification, use the `nodeAffinity` stanza to configure the `preferredDuringSchedulingIgnoredDuringExecution` parameter:
++
+.. Specify a weight for the node, 1-100. The node with highest weight is preferred.
```
May be "Specify a weight for the node, as a number between 1-100."

gaurav-nelson commented on admin_guide/scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,364 @@
+* Pod affinity can tell the scheduler to locate a new pod on the same node as other pods if the label selector on the new pod matches the label on the current pod.
+* Pod anti-affinity can prevent the scheduler from locating a new pod on the same node as pods with the same labels if the label selector on the new pod matches the label on the current pod.
+
+There are currently two types of pod affinity rules: _required_ and _preferred_.
```
Remove _Currently_: s/There are currently two/There are two

gaurav-nelson commented on admin_guide/scheduling/scheduler-advanced.adoc:
```diff
@@ -0,0 +1,53 @@
+See xref:../../admin_guide/advanced_scheduling/pod_affinity.adoc#admin-guide-sched-pod-affinity[Using Pod Affinity and Anti-Affinity].
+
+[[admin-guide-scheduling-advanced-node]]
+Node Affinity:: Node affinity allows a *pod* to specify an affinity (or anti-affinity) towards a group of *nodes* (due to their special hardware, location, requirements for high availability, and so forth) it can be placed on.  The node does not have control over the placement. 
```
Remove double space _"placed on.  The node does"_

gaurav-nelson commented on admin_guide/scheduling/scheduler-advanced.adoc:
```diff
@@ -0,0 +1,53 @@
+
+== Using Multiple Methods
+
+These different methods can be used together to create the most effificent processes for your environment. For example, if you have a dedicated group of nodes and want specific pods to be placed only on those nodes, you can uses a combination of taints/tolerations and node affinity to make sure that appropriate new pods are scheduled on those nodes and prevent other new pods from being scheduled  ...
```
s/make sure/verify (Appear in other edited documents as well.)

wmengRH replied:
effificent
efficient?

mburke5678 replied:
@gaurav-nelson @wmengRH "Verify" is not correct here. But, "make sure" might not be the best wording. It should be "to ensure that appropriate new pods are scheduled..."
I used this style guide as a reference (see "ensure" in Chapter 13),

gaurav-nelson replied:
@mburke5678 Yes, that's correct, _Ensure_ is better.

gaurav-nelson commented:
Few more comments from me. Thank you @mburke5678

mburke5678 commented:
@wmengRH @gaurav-nelson @bfallonf Is this one good to merge?

vikram-redhat commented on _topic_map.yml:
```diff
@@ -531,10 +531,27 @@ Topics:
-  - Name: Scheduler
-    File: scheduler
+    Distros: openshift-*    
+  - Name: Advacnced Scheduling
```
Also - I am wondering about the organization of this. This section now contains the original default scheduler and the new advanced scheduler and custom scheduler, but the overall section is called Advanced Scheduler. That is confusing. Any reason why we can't just leave it as Scheduler and include the other topics in it?

mburke5678 replied:
@aheslin I have moved these topics around so many times (maybe the cause of the following bad links!) that a second opinion on the organization would be most appreciated.

vikram-redhat commented on admin_guide/scheduling/scheduler-advanced.adoc:
```diff
@@ -0,0 +1,53 @@
++
+Pod affinity uses labels on nodes and label selectors on pods to create rules for pod placement. Rules can be mandatory (required) or best-effort (preferred).
++
+See xref:../../admin_guide/advanced_scheduling/pod_affinity.adoc#admin-guide-sched-pod-affinity[Using Pod Affinity and Anti-Affinity].
```
This link is not correct.

vikram-redhat commented on admin_guide/scheduling/scheduler-advanced.adoc:
```diff
@@ -0,0 +1,53 @@
++
+Node affinity uses labels on nodes and label selectors on pods to create rules for pod placement. Rules can be mandatory (required) or best-effort (preferred).
++
+See xref:../../admin_guide/advanced_scheduling/node_affinity.adoc#admin-guide-sched-affinity[Using Node Affinity].
```
Link needs fixing.

vikram-redhat commented on admin_guide/scheduling/scheduler-advanced.adoc:
```diff
@@ -0,0 +1,53 @@
+[[admin-guide-scheduling-advanced-selector]]
+Node Selectors:: Node selectors are the simplest form of advanced scheduling. Like node affinity, it also uses labels on nodes and label selectors on pods to allows a *pod* to control the *nodes* on which is can be placed. However, node selectors do not have required and preferred rules that node affinities have.
++
+See xref:../../admin_guide/advanced_scheduling/node_selector.adoc#admin-guide-sched-selector[Using Node Selectors].
```
Link not correct.

vikram-redhat commented on admin_guide/scheduling/scheduler-advanced.adoc:
```diff
@@ -0,0 +1,53 @@
++
+Taints/tolerations have one advantage over affinities. For example, if you add to a cluster a new group of nodes with different labels, you would need to update affinities on each of the pods you want to access the node and on any other pods you do not want to use the new nodes. With taints/tolerations, you would only need to update those pods that are required to land on those new nodes, because ...
++
+See xref:../../admin_guide/advanced_scheduling/taints_tolerations.adoc#admin-guide-taints[Using Taints and Tolerations].
```
Fix link.

vikram-redhat commented on admin_guide/scheduling/custom_scheduler.adoc:
```diff
@@ -0,0 +1,102 @@
+  containers:
+  - name: pod-with-second-annotation-container
+    image: gcr.io/google_containers/pause:2.0
+----
```
Callout formatting is wrong here, which is incorrectly rendering this doc. It should be:

====
----
Source
----
<1>
====

vikram-redhat commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,288 @@
+
+== Overview
+
+_Node affinity_ is a set of rules used by the scheduler to determine where a pod can be placed. The rules are defined using custom xref:../architecture/core_concepts/pods_and_services.adoc#labels[labels on nodes] and label selectors specified in pods. Node affinity allows a *pod* to specify an affinity (or anti-affinity) towards a group of *nodes* it can be placed on. The node does not have contr ...
```
Link is not correct.

vikram-redhat commented on admin_guide/scheduling/node_affinity.adoc:
```diff
@@ -0,0 +1,288 @@
+[[admin-guide-sched-affinity-selectors]]
+== Node Affinity and Node Selectors
+
+If you are using node affinity and xref:../../admin_guide/advanced_scheduling/node_selector#admin-guide-sched-selector[node selectors] in the same pod configuration, note the following:
```
Link is not correct.

vikram-redhat commented on admin_guide/scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,364 @@
+[[admin-guide-sched-affinity-pod-config]]
+== Configuring Pod Affinity/Anti-Affinity
+
+You configure pod affinity/anti-affinity through the pod specification files. You can specify a xref:admin-guide-sched-affinity-config-req[required rule], a xref:admin-guide-sched-affinity-config-pref[preferred rule], or both. If you specify both, the node must first meet the required rule, then will attempt to meet the preferred rule.
```
Invalid links

vikram-redhat commented on admin_guide/scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,364 @@
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+  containers:
+  - name: with-pod-affinity
+    image: gcr.io/google_containers/pause:2.0
```
Could we use an example other than Google's?

vikram-redhat commented on _topic_map.yml:
```diff
@@ -531,10 +531,27 @@ Topics:
+        File: custom_scheduler
+      - Name: Using Node Affinity
+        File: node_affinity
+      - Name: Using Pod Affinity and Anti-Affinity
```
Anti-affinity?

vikram-redhat commented on admin_guide/manage_nodes.adoc:
```diff
@@ -309,10 +309,10 @@ pod anti-affinity. This is an alpha feature in Kubernetes that is available for
 
 [[pod-anti-affinity]]
-=== Using Pod Anti-Affinity for the Docker Registry Pod
+=== Using Pod Anti-Affinity
```
Anti-affinity?

vikram-redhat commented on admin_guide/scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,364 @@
+    image: docker.io/ocpqe/hello-pod
+----
+
+. When creating other pods, edit the pod specification to set the following parameters:
```
Is this missing a list of parameters?

vikram-redhat commented on admin_guide/scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,364 @@
+[[admin-guide-sched-affinity-examples]]
+== Examples
+
+The following examples demonstrate pod affinity and pod anti-affinity.
```
This might need more or different explanation. If we are only now seeing examples of pod affinity and anti-affinity, what did the examples before this show?

mburke5678 replied:
The examples here are to demonstrate the full flow of pod affinity with specific values and the results of creating pods with affinity and anti-affinity.
The previous sections showed generically how to configure pod for affinity/anti-affinity.

vikram-redhat commented on admin_guide/scheduling/scheduler-advanced.adoc:
```diff
@@ -0,0 +1,53 @@
+See xref:../../admin_guide/advanced_scheduling/node_affinity.adoc#admin-guide-sched-affinity[Using Node Affinity].
+
+[[admin-guide-scheduling-advanced-selector]]
+Node Selectors:: Node selectors are the simplest form of advanced scheduling. Like node affinity, it also uses labels on nodes and label selectors on pods to allows a *pod* to control the *nodes* on which is can be placed. However, node selectors do not have required and preferred rules that node affinities have.
```
s/allows/allow

bfallonf replied:
Not sure about "Like node affinity, it also uses labels on nodes and label selectors on pods to allows a *pod* to control the *nodes* on which is can be placed." The first "it" is vague and I think breaks the tense. Maybe:

"Like node affinity, node selectors use labels on nodes and label selectors on pods to allow a *pod* to control the *nodes* on which it can be placed."

Note the s/is/it as well.

bfallonf commented on admin_guide/scheduling/scheduler-advanced.adoc:
```diff
@@ -0,0 +1,53 @@
+
+== Using Multiple Methods
+
+These different methods can be used together to create the most efficient processes for your environment. For example, if you have a dedicated group of nodes and want specific pods to be placed only on those nodes, to ensure that appropriate new pods are scheduled on those nodes and prevent other new pods from being scheduled on those nodes.
```
I don't think the section above is in the right place. Which methods are we talking about here? Why is it in its own section, but talking as thought the reader should have prior knowledge? I'd get rid of the heading as a section.

But also, I think all of the above can be changed to:

"To ensure that appropriate new pods are scheduled on a dedicated group of nodes and prevent other new pods from being scheduled on those nodes, you can combine the above methods."

That is of course, if that's actually what the paragraph is trying to say.

vikram-redhat commented on admin_guide/scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,398 @@
+* The pod has the following tolerations:
++
+----
+[source, yaml]
```
[source, yaml] in incorrect place.

mburke5678 replied:
@vikram-redhat What does this tag do and when do we use it? I see some references in public asciidoc manuals, but nothing in our team material that I found.

vikram-redhat commented on admin_guide/scheduling/pod_affinity.adoc:
```diff
@@ -0,0 +1,364 @@
+.Sample pod config file with pod affinity
+
+----
+[source,yaml]
```
[source, yaml] in the wrong place.

vikram-redhat commented on admin_guide/scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,398 @@
+.Sample pod configuration file with `Equal` operator
+
+----
+[source, yaml]
```
[source, yaml] in the wrong place.

vikram-redhat commented on admin_guide/scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,398 @@
+.Sample pod configuration file with `Exists` operator
+
+----
+[source, yaml]
```
[source, yaml] in the wrong place.

vikram-redhat commented on admin_guide/scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,398 @@
+For example:
+
+----
+[source, yaml]
```
[source, yaml] in the wrong place.

vikram-redhat commented on admin_guide/scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,398 @@
+. Add a toleration to pods:
++
+----
+[source, yaml]
```
[source, yaml] in the wrong place.

vikram-redhat commented on admin_guide/scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,398 @@
+or
++
+----
+[source, yaml]
```
[source, yaml] in the wrong place.

vikram-redhat commented on admin_guide/scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,398 @@
+$ oadm taint nodes node1 dedicated=groupName:NoSchedule
+----
+
+. Add a corresponding toleration to the pods by writing a custom xref:../install_config/master_node_configuration.adoc#master-config-admission-control-config[admission controller].
```
Invalid link

vikram-redhat commented on admin_guide/scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,398 @@
+$ oadm taint nodes node1 dedicated=groupName:NoSchedule
+----
+
+. Add a corresponding toleration to the pods by writing a custom xref:../install_config/master_node_configuration.adoc#master-config-admission-control-config[admission controller].
```
Fix link

vikram-redhat commented on admin_guide/scheduling/taints_tolerations.adoc:
```diff
@@ -0,0 +1,398 @@
+$ oadm taint nodes <node-name> disktype=ssd:PreferNoSchedule
+----
+
+. Adding a corresponding toleration to pods that use the special hardware using an xref:../../architecture/additional_concepts/admission_controllers.adoc[admission controller].
```
Link is not correct.

mburke5678 commented:
@aveshagarwal   I added a new section to the Default Scheduler topic, "Control Pod Placement to Projects," based on Derek's comment that this Trello is related to scheduling.
admin_guide/scheduling/scheduler.adoc 
PTAL

aveshagarwal commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -0,0 +1,819 @@
+      configuration:
+        apiversion: v1
+        kind:podNodeSelectorPluginConfig: <1>
+          clusterDefaultNodeSelector: region=west <2>
```
Could you provide some more information about clusterDefaultNodeSelector? Something like that if a project does not have any node selectors specified, pods for that project will be merged with clusterDefaultNodeSelector.

aveshagarwal commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -0,0 +1,819 @@
+        apiversion: v1
+        kind:podNodeSelectorPluginConfig: <1>
+          clusterDefaultNodeSelector: region=west <2>
+          ns1: region=west,env=test,infra=fedora,os=fedora
```
Could you clearly specify that this acts as a whitelist?

mburke5678 commented:
@aveshagarwal Can you take a quick look to make sure I have addressed your concerns so that we can move the docs Trello to complete and start the QE review?

aveshagarwal commented:
@mburke5678 lgtm now. Thanks.

aveshagarwal commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -0,0 +1,827 @@
+    PodNodeSelector:
+      configuration:
+        apiversion: v1
+        kind:podNodeSelectorPluginConfig: <1>
```
@mburke5678 I realized this configuration is not right for this plugin pod node selector. Right way to configure this is:


```
admissionConfig:
  pluginConfig:
    PodNodeSelector:
      configuration:
        podNodeSelectorPluginConfig:
          clusterDefaultNodeSelector: "k3=v3"
          ns1: region=west,env=test,infra=fedora,os=fedora
```

Or 

```
admissionConfig:
  pluginConfig:
    PodNodeSelector:
      location: <path-to-file>
```

Where the file will look like:

```
podNodeSelectorPluginConfig:
    clusterDefaultNodeSelector: "k3=v3"
     ns1: region=west,env=test,infra=fedora,os=fedora
```

wmengRH commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -0,0 +1,827 @@
+apiVersion: v1
+kind: Pod
+spec:
+  nodeName: <key: value>
```
I think here should be only a value, not key:value pair. like:
nodeName: hostxxxx.redhat.com
@aveshagarwal correct me if I am wrong

aveshagarwal replied:
@wmengRH correct.

mburke5678 replied:
@aveshagarwal @wmengRH 

Is that a literal string? 'nodeName: hostxxxx.redhat.com`

aveshagarwal replied:
nodeName is a field, and it value is a string.

mburke5678 replied:
@aveshagarwal @wmengRH 

So, replace with:
nodeName: <value>

wmengRH commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -0,0 +1,827 @@
+----
+
+If there is a node that does not have a node selector annotation, the pod will be scheduled there. If there is no node
+
```
seems not finished.
"If there is a node that does not have a node selector annotation, the pod will be scheduled there. If there is no node"

mburke5678 commented:
@openshift/team-documentation 
I added a new section to the Default Scheduler topic, "Control Pod Placement to Projects."
admin_guide/scheduling/scheduler.adoc
PTAL

ahardin-rh commented:
@mburke5678 This is a biggie! Can you please provide a preview build?

mburke5678 commented:
@ahardin-rh All of the other sections have been peer reviewed by Brice, Gaurav, and Vikram except for a new section I added to the Default Scheduler topic, called "Control Pod Placement to Projects."
http://file.rdu.redhat.com/mburke/trello-node-affinity/admin_guide/scheduling/scheduler.html#control-pod-placement-to-projects

That is the part I need peer-reviewed. But, feel free to look at any and all other sections! Thanks!!

mburke5678 commented:
@openshift/team-documentation 
All of the other sections have been peer reviewed by Brice, Gaurav, and Vikram except for a new section I added to the Default Scheduler topic, called "Control Pod Placement to Projects."
http://file.rdu.redhat.com/mburke/trello-node-affinity/admin_guide/scheduling/scheduler.html#control-pod-placement-to-projects

That is the part I need peer-reviewed.

ahardin-rh commented:
@mburke5678 I see the Control Pod Placement to Projects section in the preview build, but not pushed here for comments. I'll comment here instead:

    1. /etc/origin/master/master-config.yaml needs markup *_/etc/origin/master/master-config.yaml _* (I see 2 instances.)
    2. s/ Or, /Alternatively,

mburke5678 commented:
[rev_history]
|xref:../admin_guide/scheduling/index.adoc#admin-guide-scheduling-index[Scheduling]
|Added subsections for new scheduling features. Moved the current Scheduling topic into section as xref:../admin_guide/scheduling/scheduler.adoc#admin-guide-schedulerDefault Scheduling].
%

vikram-redhat commented:
@mburke5678 - lots of duplicated ids all throughout. Please be more careful. I have fixed it for this release.
